# Car Price Prediction (Used Cars)

In [1]:
#import data from sklearn built-in datasets
import pandas as pd
import warnings
warnings.filterwarnings('ignore')
df=pd.read_csv('dataset/car data.csv',usecols=['Car_Name','Year','Present_Price','Driven_kms','Fuel_Type','Selling_type','Transmission','Selling_Price'])

# change column position from begining to last
column_to_move = df.pop("Selling_Price")
df.insert(7, "Selling_Price", column_to_move)
df

Car_Name  Year  Present_Price  Driven_kms Fuel_Type Selling_type  \
0       ritz  2014           5.59       27000    Petrol       Dealer   
1        sx4  2013           9.54       43000    Diesel       Dealer   
2       ciaz  2017           9.85        6900    Petrol       Dealer   
3    wagon r  2011           4.15        5200    Petrol       Dealer   
4      swift  2014           6.87       42450    Diesel       Dealer   
..       ...   ...            ...         ...       ...          ...   
296     city  2016          11.60       33988    Diesel       Dealer   
297     brio  2015           5.90       60000    Petrol       Dealer   
298     city  2009          11.00       87934    Petrol       Dealer   
299     city  2017          12.50        9000    Diesel       Dealer   
300     brio  2016           5.90        5464    Petrol       Dealer   

    Transmission  Selling_Price  
0         Manual           3.35  
1         Manual           4.75  
2         Manual           7.25  
3         Manual           2.85  
4         Manual           4.60  
..           ...            ...  
296       Manual           9.50  
297       Manual           4.00  
298       Manual           3.35  
299       Manual          11.50  
300       Manual           5.30  

[301 rows x 8 columns]

In [2]:
df.Car_Name.value_counts()

Car_Name
city                        26
corolla altis               16
verna                       14
fortuner                    11
brio                        10
                            ..
Honda CB Trigger             1
Yamaha FZ S                  1
Bajaj Pulsar 135 LS          1
Activa 4g                    1
Bajaj Avenger Street 220     1
Name: count, Length: 98, dtype: int64

We have total 98 different cars🚗..

In [3]:
df.Year.value_counts().sort_index()

Year
2003     2
2004     1
2005     4
2006     4
2007     2
2008     7
2009     6
2010    15
2011    19
2012    23
2013    33
2014    38
2015    61
2016    50
2017    35
2018     1
Name: count, dtype: int64

😐It has much older cars ..

In [4]:
df.Fuel_Type.value_counts()

Fuel_Type
Petrol    239
Diesel     60
CNG         2
Name: count, dtype: int64

hmmmm...239 petrol cars, 60 diesel cars and 2 CNG cars. CNG should have more..

In [5]:
df.Selling_type.value_counts()

Selling_type
Dealer        195
Individual    106
Name: count, dtype: int64

We can directly buy car from individual person from 106 people and 195 dealer present here to buy from.

In [6]:
df.Transmission.value_counts()

Transmission
Manual       261
Automatic     40
Name: count, dtype: int64

In [7]:
df.describe()

Year  Present_Price     Driven_kms  Selling_Price
count   301.000000     301.000000     301.000000     301.000000
mean   2013.627907       7.628472   36947.205980       4.661296
std       2.891554       8.642584   38886.883882       5.082812
min    2003.000000       0.320000     500.000000       0.100000
25%    2012.000000       1.200000   15000.000000       0.900000
50%    2014.000000       6.400000   32000.000000       3.600000
75%    2016.000000       9.900000   48767.000000       6.000000
max    2018.000000      92.600000  500000.000000      35.000000

In [8]:
df.duplicated().sum()

2

There are 2 duplicate values. Let's just remove it.

In [9]:
df=df.drop_duplicates(ignore_index=True)
df.duplicated().sum()

0

In [10]:
df.corr(numeric_only=True)

Year  Present_Price  Driven_kms  Selling_Price
Year           1.000000      -0.053167   -0.525714       0.234369
Present_Price -0.053167       1.000000    0.205224       0.876305
Driven_kms    -0.525714       0.205224    1.000000       0.028566
Selling_Price  0.234369       0.876305    0.028566       1.000000

Driven_kms and Year are highly corrilated inversely, meaning as Year increases Driven_kms decreased or Driven_kms increases Year decreases.

Now most people sell their cars without riding it most. But before in time, people rides a lot and then sell it.

In [11]:
df.isna().sum()

Car_Name         0
Year             0
Present_Price    0
Driven_kms       0
Fuel_Type        0
Selling_type     0
Transmission     0
Selling_Price    0
dtype: int64

No null values in the dataset.

In [12]:
#transforming categorical column as numerical column
cat_col=['Fuel_Type','Selling_type','Transmission']
i=0
for col in cat_col:
    print(f'{col}')
    for value in df[col].value_counts().index:
        df[col]=df[col].replace(to_replace=value,value=i)
        i+=1
        print(f'     {value}={i}')

Fuel_Type
     Petrol=1
     Diesel=2
     CNG=3
Selling_type
     Dealer=4
     Individual=5
Transmission
     Manual=6
     Automatic=7


In [13]:
df

Car_Name  Year  Present_Price  Driven_kms  Fuel_Type  Selling_type  \
0       ritz  2014           5.59       27000          0             3   
1        sx4  2013           9.54       43000          1             3   
2       ciaz  2017           9.85        6900          0             3   
3    wagon r  2011           4.15        5200          0             3   
4      swift  2014           6.87       42450          1             3   
..       ...   ...            ...         ...        ...           ...   
294     city  2016          11.60       33988          1             3   
295     brio  2015           5.90       60000          0             3   
296     city  2009          11.00       87934          0             3   
297     city  2017          12.50        9000          1             3   
298     brio  2016           5.90        5464          0             3   

     Transmission  Selling_Price  
0               5           3.35  
1               5           4.75  
2               5           7.25  
3               5           2.85  
4               5           4.60  
..            ...            ...  
294             5           9.50  
295             5           4.00  
296             5           3.35  
297             5          11.50  
298             5           5.30  

[299 rows x 8 columns]

In [14]:
#function to check outliers
import numpy as np
def detect_outliers(data):
    outliers=[]
    threshold=3
    mean = np.mean(data)
    std =np.std(data)
    for i in data:
        z_score= (i - mean)/std
        if np.abs(z_score) > threshold:
            outliers.append(i)
    return outliers
i=0
for feature in df.columns:
    if i!=0:   #because the first column is categorical 
        print(f'Outlier in {feature} = {detect_outliers(df[feature])}')
    i=1

Outlier in Year = [2003, 2003, 2004]
Outlier in Present_Price = [35.96, 35.96, 35.96, 36.23, 92.6]
Outlier in Driven_kms = [197176, 213000, 500000]
Outlier in Fuel_Type = [2, 2]
Outlier in Selling_type = []
Outlier in Transmission = []
Outlier in Selling_Price = [23.0, 19.99, 23.5, 33.0, 19.75, 23.0, 35.0, 20.75]


It seems that there are a lot of outliers. Lets just ignore them.

-----

In [15]:
# train test split
from sklearn.model_selection import train_test_split
X=df.iloc[:,1:-1].values
y=df.iloc[:,-1].values
X_train,X_test,y_train,y_test=train_test_split(X,y,random_state=0) #stratify is used for consistent data dividation 

In [16]:
#scaling (min-max)
from sklearn.preprocessing import MinMaxScaler
scaler=MinMaxScaler()
X_train=scaler.fit_transform(X_train)

In [17]:
#train the model with KNeighborsClassifier
from sklearn.linear_model import LinearRegression
model=LinearRegression()
model.fit(X_train,y_train)

LinearRegression()

In [18]:
model.score(X_train,y_train)

0.892106503279255

In [19]:
#r2 score
from sklearn.metrics import r2_score
pred=model.predict(X_train)
r2_score(y_train,pred)

0.892106503279255

In [20]:
#scoring on testing data
X_test=scaler.transform(X_test)
from sklearn.metrics import r2_score
pred=model.predict(X_test)
r2_score(y_test,pred)

0.8002247173994053

Finally model is trained and score of model is 89% on seen data and on unseen data the socore is 80%.